File: Python/ExampleTests/common/filter_value_string.py
from . import tokens
DNC_STRING = '?DNC?'


class FilterValueString:
    def __init__(self, value: str = ''):
        self.value = value

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 1, 'FilterValue')
        return cls(
            parts[0]
        )

    def __str__(self):
        return (tokens.token(self.value))

    def _key(self):
        return (self.value,)

    def __eq__(self, other):
        if not isinstance(other, FilterValueString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'FilterValueString({self})'

File: Swift/ExampleTests/common/FilterValueString.swift
public struct FilterValueString: CustomStringConvertible, Equatable {
    public let value: String

    public init(value: String) {
        self.value = value
    }

    public init(fromArray v: [String]) {
        self.value = v.count > 0 ? v[0] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> FilterValueString {
        let parts = Tokens.require(text, 1, "FilterValue")
        return FilterValueString(value: parts[0])
    }

    public var description: String {
        return Tokens.token(value)
    }

    public static let dncString = "?DNC?"

    public static func == (a: FilterValueString, b: FilterValueString) -> Bool {
        return (a.value == dncString || b.value == dncString || a.value == b.value)
    }
}
